README.md
MyTAP 1.0
==========

MyTAP is a unit testing framework for MySQL 5.x written using fuctions and
procedures. It includes a collection of TAP-emitting assertion functions, as
well as the ability to integrate with other TAP-emitting test frameworks.

Installation
============

To install MyTAP to a MySQL server, just run the install script from the mytap directory:

    # ./install.sh

This assumes local server access or user and host credentials in ~/.my.cnf.

The installer will create a database called `tap` with 2 tables, import the
base package and any version specific patches. When installation has completed,
the test suite will run automatically. Any errors at this point should be
raised as issues because it will likely indicate a change in the MySQL
`information_schema` or a bug in a MySQL release against which myTAP has not
been tested.

For installations on remote servers, non-standard port numbers, or where client
credentials are not stored, provide the appropriate short or long-form switches
for the required parameters. e.g. 
    
    # ./install.sh --port 3310
    # ./install.sh --user root --password rootpassword 2>/dev/null
    # ./install.sh -h 127.0.0.1 -P 3310 --u root --p rootpassword 2>/dev/null


The installer can run a subset of the tests with the --filter option

    # ./install.sh --filter table

 or, to skip the tests entirely

    # ./install.sh --no-tests

Finally, it is also possible to run the installer in test-only mode (useful for
developers) with

    # ./install.sh --no-install

Short-form versions of all of these switches exist, for full details

    # ./install.sh --help


MyTAP Test Scripts
==================

Here's an example of how to write a MyTAP test script:

    -- Start a transaction.
    BEGIN;

    -- Plan the tests.
    SELECT tap.plan(1);

    -- Run the tests.
    SELECT tap.pass( 'My test passed, w00t!' );

    -- Finish the tests and clean up.
    CALL tap.finish();
    ROLLBACK;

Note how the TAP test functions are reference from another database so as to
keep them separate from your application database.

Now you're ready to run your test script!

    % mysql -u root --disable-pager --batch --raw --skip-column-names --unbuffered --database test --execute 'source test.sql'
    1..1
    ok 1 - My test passed, w00t!

Yeah, that's rather a lot of options to have to remember to get valid tap. I
suggest that you install
[TAP::Parser::SourceHandler::MyTAP](http://search.cpan.org/dist/TAP-Parser-SourceHandler-MyTAP)
instead and just use its [`my_prove`](http://search.cpan.org/perldoc?my_prove)
utility:

    % my_prove -u root --database test test.sql

Using MyTAP
===========

The purpose of MyTAP is to provide a wide range of testing utilities that
output TAP. TAP, or the "Test Anything Protocol", is an emerging standard for
representing the output from unit tests. It owes its success to its format as
a simple text-based interface that allows for practical machine parsing and
high legibility for humans. TAP started life as part of the test harness for
Perl but now has implementations in C/C++, Python, PHP, JavaScript, Perl,
PostgreSQL, and now MySQL.

More information on the use of MyTAP can be found in the [documentation](https://hepabolu.github.io/mytap).


To Do
-----
* Port lot of other assertion functions from [pgTAP](http://pgtap.org/).

Public Repository
-----------------

The source code for MyTAP is available on
[GitHub](http://github.com/hepabolu/mytap/). Please feel free to fork and
contribute!

Authors
------

* [David E. Wheeler](http://justatheory.com/)
* [Hepabolu](https://github.com/hepabolu/mytap)
* [Paul Campbell](https://github.com/animalcarpet)

Credits
-------

* Michael Schwern and chromatic for Test::More.
* Adrian Howard for Test::Exception.

Copyright and License
---------------------

Copyright (c) 2010 David E. Wheeler, Helma van der Linden. Some rights reserved.

The full license is available in a separate LICENSE file.



Changes
Revision history for MyTAP
==========================

1.0   2018-07-21
-------------------------
This is a massive rewrite and extension of the available tests with thanks to @animalcarpet.
This version also has version specific tests and uses slightly different tests based
on  the MySQL version.
This version is tested on MySQL 5.5, 5.6 and 5.7

0.07  2018-06-18
-------------------------
* col_has_unique_index, col_hasnt_unique_index added to test for
  the existence or absence of a unique index.
* tests added for the new functions.
* syntax errors fixed in the tests for the has[nt]_function and
  has[nt]_procedure functions.
* runtests.sh script now has a filter option.

0.06  2017-06-18
-------------------------
* has_table and hasnt_table moved to separate file
* has_trigger and hasnt_trigger added to test if a table has a
  certain trigger
* tests added for the new functions

0.05  2016-06-22T19:25:12
-------------------------
* col_has_pos_in_named_index, col_has_pos_in_named_index added to test
  for index keys that combine multiple columns.
* tests added for the new functions.

0.04  2016-01-01T21:55:58
-------------------------
* col_has_named_index, col_hasnt_named_index added to allow index key names
  that differ from the column name. Defaults to column name.
* tests added for the new functions

0.03  2015-04-26T08:12:37
-------------------------
* _has_table reworked to distinguish between views and tables
* has_table and hasnt_tabel reworked to allow custom description
* has_view, hasnt_view, has_security_invoker, has_security_definer tests added
* all kinds of column related tests added
* has_function, hasnt_function, has_procedure, hasnt_procedure added

Note: these tests are only tested on MySQL 5.5x

0.02  2010-08-30T17:36:29
-------------------------
* Modified exception-throwing code to always clear out the cached data before
  dying. Doesn't cover cases where the code being tested dies, but at least it
  covers our own intentional deaths.
* Fixed `DROP` lines to be terminated with "//" instead of ";". Seems like
  some versions of MySQl didn't mind the semicolon, but others did.


0.01  2010-07-28T18:37:03
-------------------------
* Initial public release with core functionality ported from
  [pgTAP](http://pgtap.org/).
  [Announcement](http://justatheory.com/computers/databases/mysql/introducing_mytap.html).


runtests.sh
#!/bin/bash

SQLHOST=localhost;
SQLPORT=3306;

while [[ "$#" > 0 ]]; do
    case $1 in
	-u|--user)
	    SQLUSER="$2";
	    shift
	    ;;
	-p|--password)
	    SQLPASS="$2"
	    shift
	    ;;
	-h|--host)
	    SQLHOST="$2"
	    shift
	    ;;
	-P|--port)
	    SQLPORT="$2"
	    shift
	    ;;
	-f|--filter)
	    FILTER="$2"
	    shift
	    ;;
	-h|--help)
	    cat << EOF
Usage:
 runtests.sh [options]

Options:
 -u, --user           MySQL username
 -p, --password       MySQL password
 -P, --port           MySQL port
 -h, --host           MySQL host
 -f, --filter         <matching|eq|moretap|todo|utils|charset|collation|column|constraint|engine|event|index|partition|routines|table|trigger|schemata|user|view>
EOF
	   exit 1 
	   ;;
	 *)     
	   exit 1
	   ;;
    esac;
    shift;
done


MYSQLOPTS="--disable-pager --batch --raw --skip-column-names --unbuffered"

if [[ $SQLUSER != '' ]] && [[ $SQLPASS != '' ]]; then
  MYSQLOPTS="$MYSQLOPTS -u$SQLUSER -p$SQLPASS"
fi

if [[ $SQLHOST != 'localhost' ]]; then
  MYSQLOPTS="$MYSQLOPTS --host $SQLHOST"
fi

if [[ $SQLPORT != '3306' ]]; then
  MYSQLOPTS="$MYSQLOPTS --port $SQLPORT"
fi

if [[ $FILTER == '' ]]; then
  FILTER=0
fi

MYVER1=`mysql $MYSQLOPTS --execute "SELECT @@global.version" | awk -F'-' '{print $1}' | awk -F'.' '{print $1 * 100000 }'`;
MYVER2=`mysql $MYSQLOPTS --execute "SELECT @@global.version" | awk -F'-' '{print $1}' | awk -F'.' '{print $2 * 1000 }'`;
MYVER3=`mysql $MYSQLOPTS --execute "SELECT @@global.version" | awk -F'-' '{print $1}' | awk -F'.' '{print $3}'`;


MYVER=$(($MYVER1 + $MYVER2 + $MYVER3));

# import the full package before running the tests
# you can't use a wildcard with the source command so all version specific files need
# to be separately listed

echo "============= updating tap ============="
echo "Importing myTAP base"
mysql $MYSQLOPTS --execute 'source ./mytap.sql'
mysql $MYSQLOPTS --execute 'source ./mytap-global.sql';

if [[ $MYVER -gt 506000 ]]; then
    echo "Importing Version 5.6 patches";
    mysql $MYSQLOPTS --execute 'source ./mytap-table-56.sql';
fi

if [[ $MYVER -gt 507000 ]]; then
    echo "Importing Version 5.7 patches";
    mysql $MYSQLOPTS --execute 'source ./mytap-table-57.sql';
fi

if [[ $MYVER -gt 507006 ]]; then
    echo "Importing Version 5.7.6 patches";
    mysql $MYSQLOPTS --execute 'source ./mytap-global-57.sql';
fi

if [[ $MYVER -gt 800000 ]]; then
    echo "Importing Version 8.0 patches";
    mysql $MYSQLOPTS --execute 'source ./mytap-table-80.sql';
fi

if [[ $FILTER != 0 ]]; then
  echo "============= filtering ============="
  echo "$FILTER"
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "matching" ]]; then
  echo "============= matching ============="
  mysql $MYSQLOPTS --database tap --execute 'source tests/matching.my'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "eq" ]]; then
  echo "============= eq ============="
  mysql $MYSQLOPTS --database tap --execute 'source tests/eq.my'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "moretap" ]]; then
  echo "============= moretap ============="
  mysql $MYSQLOPTS --database tap --execute 'source tests/moretap.my'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "todotap" ]]; then
  echo "============= todotap ============="
  mysql $MYSQLOPTS --database tap --execute 'source tests/todotap.my'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "utils" ]]; then
  echo "============= utils ============="
  mysql $MYSQLOPTS --database tap --execute 'source tests/utils.my'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "charset" ]]; then
  echo "============= character sets ============"
  mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-charset.sql'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "collation" ]]; then
  echo "============= collations ============"
  mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-collation.sql'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "column" ]]; then
  echo "============= columns ============"
  mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-column.sql'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "constraint" ]]; then
  echo "============= cconstraints ============"
  mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-constraint.sql'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "engine" ]]; then
  echo "============= engines ============"
  mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-engine.sql'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "event" ]]; then
  echo "============= events ============"
  mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-event.sql'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "index" ]]; then
  echo "============= indexes ============"
  mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-index.sql'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "partition" ]]; then
  echo "============= partitions ============"
  mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-partition.sql'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "routines" ]]; then
  echo "============= routines ============"
  mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-routines.sql'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "schemata" ]]; then
  echo "============= schemas ============"
  mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-schemata.sql'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "table" ]]; then
  echo "============= tables ============"
  mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-table.sql'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "trigger" ]]; then
  echo "============= triggers ============"
  mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-trigger.sql'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "user" ]]; then
  echo "============= users ============"
  mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-user.sql'
fi

if [[ $FILTER == 0 ]] || [[ $FILTER =~ "view" ]]; then
  echo "============= views ============"
  mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-view.sql'
fi



runsandboxtests.sh
#!/bin/bash
#
# Run tests against the various sandboxed MySQL servers


USER=msandbox
PW=msandbox
HOST=127.0.0.1

PORT55=3306
PORT56=5636
PORT57=5718
PORT80=8001

MYSQLOPTS="-h $HOST -u $USER -p$PW"

# ==== MySQL 5.5

echo "============= updating tap in 5.5 ============="
mysql $MYSQLOPTS --port=$PORT55 --execute 'source ./mytap.sql'

myprove/bin/my_prove tests/* -h $HOST -P $PORT55 -u $USER -p $PW

# ==== MySQL 5.6

echo "============= updating tap in 5.6 ============="
mysql $MYSQLOPTS --port=$PORT56 --execute 'source ./mytap.sql'

myprove/bin/my_prove tests/* -h $HOST -P $PORT56 -u $USER -p $PW

# ==== MySQL 5.7

echo "============= updating tap in 5.7 ============="
mysql $MYSQLOPTS --port=$PORT57 --execute 'source ./mytap.sql'

myprove/bin/my_prove tests/* -h $HOST -P $PORT57 -u $USER -p $PW

# ==== MySQL 8.0

# echo "============= updating tap in 8.0 ============="
# mysql $MYSQLOPTS --port=$PORT80 --execute 'source ./mytap.sql'

# myprove/bin/my_prove tests/* -h $HOST -P $PORT80 -u $USER -p $PW



assets/js/jquery.toc.js
/*
 * jQuery Table of Content Generator for Markdown v1.0
 *
 * https://github.com/dafi/tocmd-generator
 * Examples and documentation at: https://github.com/dafi/tocmd-generator
 *
 * Requires: jQuery v1.7+
 *
 * Copyright (c) 2013 Davide Ficano
 *
 * Dual licensed under the MIT and GPL licenses:
 *   http://www.opensource.org/licenses/mit-license.php
 *   http://www.gnu.org/licenses/gpl.html
 */
(function($) {
    var toggleHTML = '<div id="toctitle">' +
    '<span class="toctoggle"><a id="toctogglelink" class="internal" href="#">%2</a></span>  <h2>%1</h2>' +
    '</div>';
    var tocContainerHTML = '<div id="toc-container">' +
    '<div class="toc" id="toc">' +
    '%1<ul>%2</ul>' +
    '</div></div>';

    function createLevelHTML(anchorId, tocLevel, tocSection, tocNumber, tocText, tocInner) {
        var link = '<a href="#%1"><span class="tocnumber">%2</span> <span class="toctext">%3</span></a>%4'
            .replace('%1', anchorId)
            .replace('%2', tocNumber)
            .replace('%3', tocText)
            .replace('%4', tocInner ? tocInner : '');
        return '<li class="toclevel-%1 tocsection-%2">%3</li>\n'
            .replace('%1', tocLevel)
            .replace('%2', tocSection)
            .replace('%3', link);
    }

    $.fn.toc = function(settings) {
        var config = {
            renderIn: 'self',
            anchorPrefix: 'tocAnchor-',
            showAlways: false,
            minItemsToShowToc: 2,
            saveShowStatus: true,
            contentsText: 'Contents',
            hideText: 'hide',
            showText: 'show',
            showCollapsed: false};

        if (settings) {
            $.extend(config, settings);
        }

        var tocHTML = '';
        var tocLevel = 1;
        var tocSection = 1;
        var itemNumber = 1;

        var tocContainer = $(this);

        console.log('toccontainer', tocContainer);

        tocContainer.find('h1').each(function() {
            var levelHTML = '';
            var innerSection = 0;
            var h1 = $(this);

            console.log('h1 found: ', h1);

            h1.nextUntil('h1').filter('h2').each(function() {
                ++innerSection;
                var anchorId = config.anchorPrefix + tocLevel + '-' + tocSection + '-' +  + innerSection;
                $(this).attr('id', anchorId);
                levelHTML += createLevelHTML(anchorId,
                    tocLevel + 1,
                    tocSection + innerSection,
                    itemNumber + '.' + innerSection,
                    $(this).text());
            });
            if (levelHTML) {
                levelHTML = '<ul>' + levelHTML + '</ul>\n';
            }
            var anchorId = config.anchorPrefix + tocLevel + '-' + tocSection;
            h1.attr('id', anchorId);
            tocHTML += createLevelHTML(anchorId,
                tocLevel,
                tocSection,
                itemNumber,
                h1.text(),
                levelHTML);

            tocSection += 1 + innerSection;
            ++itemNumber;
        });

        // for convenience itemNumber starts from 1
        // so we decrement it to obtain the index count
        var tocIndexCount = itemNumber - 1;

        var show = config.showAlways ? true : config.minItemsToShowToc <= tocIndexCount;

        // check if cookie plugin is present otherwise doesn't try to save
        if (config.saveShowStatus && typeof($.cookie) == "undefined") {
            config.saveShowStatus = false;
        }

        if (show && tocHTML) {
            var replacedToggleHTML = toggleHTML
                .replace('%1', config.contentsText)
                .replace('%2', config.hideText);
            var replacedTocContainer = tocContainerHTML
                .replace('%1', replacedToggleHTML)
                .replace('%2', tocHTML);

            // Renders in default or specificed path
            if (config.renderIn != 'self') {
              $(config.renderIn).html(replacedTocContainer);
            } else {
              tocContainer.prepend(replacedTocContainer);
            }

            $('#toctogglelink').click(function() {
                var ul = $($('#toc ul')[0]);

                if (ul.is(':visible')) {
                    ul.hide();
                    $(this).html(config.showText);
                    if (config.saveShowStatus) {
                        $.cookie('toc-hide', '1', { expires: 365, path: '/' });
                    }
                    $('#toc').addClass('tochidden');
                } else {
                    ul.show();
                    $(this).html(config.hideText);
                    if (config.saveShowStatus) {
                        $.removeCookie('toc-hide', { path: '/' });
                    }
                    $('#toc').removeClass('tochidden');
                }
                return false;
            });

            if (config.saveShowStatus && $.cookie('toc-hide')) {
                var ul = $($('#toc ul')[0]);

                ul.hide();
                $('#toctogglelink').html(config.showText);
                $('#toc').addClass('tochidden');
            }

            if (config.showCollapsed) {
                $('#toctogglelink').click();
            }
        }
        return this;
    }
})(jQuery);


install.sh
#!/bin/bash

# Installation script for MyTAP


SQLHOST=localhost;
SQLPORT=3306;
NOTESTS=0
NOINSTALL=0
FILTER=0

while [[ "$#" > 0 ]]; do
    case $1 in
	-u|--user)
	    SQLUSER="$2";
	    shift
	    ;;
	-p|--password)
	    SQLPASS="$2"
	    shift
	    ;;
	-h|--host)
	    SQLHOST="$2"
	    shift
	    ;;
	-P|--port)
	    SQLPORT="$2"
	    shift
	    ;;
	-f|--filter)
	    NOFILTER=0
	    FILTER="$2"
	    shift
	    ;;
	-t|--no-tests)
	    NOTESTS=1
	    ;;
	-i|--no-install)
	    NOINSTALL=1
	    ;;
	-h|--help)
	    cat << EOF
Usage:
 install.sh [options]

Options:
 -u, --user           MySQL username
 -p, --password       MySQL password
 -P, --port           MySQL port
 -h, --host           MySQL host
 -t, --no-tests       Don't run the test suite when the install is completed
 -i, --no-install     Don't perform the installation, i.e. just run the test suite
 -f, --filter         Perform the action on one class of objects <matching|eq|moretap|todo|utils|charset|collation|column|constraint|engine|event|index|partition|routines|table|trigger|schemata|user|view>
EOF
	   exit 1 
	   ;;
	 *)     
	   exit 1
	   ;;
    esac;
    shift;
done


MYSQLOPTS="--disable-pager --batch --raw --skip-column-names --unbuffered"

if [[ $SQLUSER != '' ]] && [[ $SQLPASS != '' ]]; then
    MYSQLOPTS="$MYSQLOPTS -u$SQLUSER -p$SQLPASS";
    CMD="--user $SQLUSER --password = $SQLPASS";
fi

if [[ $SQLHOST != 'localhost' ]]; then
   MYSQLOPTS="$MYSQLOPTS --host $SQLHOST";
   CMD="$CMD --host $SQLHOST";
fi

if [[ $SQLPORT != '3306' ]]; then
  MYSQLOPTS="$MYSQLOPTS --port $SQLPORT"
  CMD="$CMD --port $SQLPORT"
fi

MYVER1=`mysql $MYSQLOPTS --execute "SELECT @@global.version" | awk -F'-' '{print $1}' | awk -F'.' '{print $1 * 100000 }'`;
MYVER2=`mysql $MYSQLOPTS --execute "SELECT @@global.version" | awk -F'-' '{print $1}' | awk -F'.' '{print $2 * 1000 }'`;
MYVER3=`mysql $MYSQLOPTS --execute "SELECT @@global.version" | awk -F'-' '{print $1}' | awk -F'.' '{print $3}'`;


MYVER=$(($MYVER1 + $MYVER2 + $MYVER3));

# import the full package before running the tests
# you can't use a wildcard with the source command so all version specific files need
# to be separately listed


if [[ $NOINSTALL -eq 0 ]]; then
    echo "============= installing myTAP ============="
    echo "Importing myTAP base"
    mysql $MYSQLOPTS --execute 'source ./mytap.sql';

    if [[ $MYVER -gt 506000 ]]; then
       echo "Importing Version 5.6 patches";
       mysql $MYSQLOPTS --execute 'source ./mytap-table-56.sql';
    fi

    if [[ $MYVER -gt 507000 ]]; then
       echo "Importing Version 5.7 patches";
       mysql $MYSQLOPTS --execute 'source ./mytap-table-57.sql';
    fi

    if [[ $MYVER -gt 507006 ]]; then
       echo "Importing Version 5.7.6 patches";
       mysql $MYSQLOPTS --execute 'source ./mytap-global-57.sql';
    fi

    if [[ $MYVER -gt 800000 ]]; then
       echo "Importing Version 8.0 patches";
       mysql $MYSQLOPTS --execute 'source ./mytap-table-80.sql';
    fi
fi

if [[ $NOTESTS -eq 0 ]]; then
   if [[ $FILTER != 0 ]]; then
      echo "Running test suite with filter: $FILTER";
   else
      echo "Running Full test suite, this will take a couple of minutes to complete."
   fi

   sleep 2;

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "matching" ]]; then
      echo "============= matching ============="
      mysql $MYSQLOPTS --database tap --execute 'source tests/matching.my'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "eq" ]]; then
      echo "============= eq ============="
      mysql $MYSQLOPTS --database tap --execute 'source tests/eq.my'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "moretap" ]]; then
      echo "============= moretap ============="
      mysql $MYSQLOPTS --database tap --execute 'source tests/moretap.my'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "todotap" ]]; then
      echo "============= todotap ============="
      mysql $MYSQLOPTS --database tap --execute 'source tests/todotap.my'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "utils" ]]; then
      echo "============= utils ============="
      mysql $MYSQLOPTS --database tap --execute 'source tests/utils.my'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "charset" ]]; then
      echo "============= character sets ============"
      mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-charset.sql'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "collation" ]]; then
      echo "============= collations ============"
      mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-collation.sql'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "column" ]]; then
      echo "============= columns ============"
      mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-column.sql'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "constraint" ]]; then
      echo "============= constraints ============"
      mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-constraint.sql'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "engine" ]]; then
      echo "============= engines ============"
      mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-engine.sql'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "event" ]]; then
      echo "============= events ============"
      mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-event.sql'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "index" ]]; then
      echo "============= indexes ============"
      mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-index.sql'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "partition" ]]; then
      echo "============= partitions ============"
      mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-partition.sql'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "routines" ]]; then
      echo "============= routines ============"
      mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-routines.sql'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "schemata" ]]; then
      echo "============= schemas ============"
      mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-schemata.sql'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "table" ]]; then
      echo "============= tables ============"
      mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-table.sql'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "trigger" ]]; then
      echo "============= triggers ============"
      mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-trigger.sql'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "user" ]]; then
      echo "============= users ============"
      mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-user.sql'
   fi

   if [[ $FILTER == 0 ]] || [[ $FILTER =~ "view" ]]; then
      echo "============= views ============"
      mysql $MYSQLOPTS --database tap --execute 'source tests/test-mytap-view.sql'
   fi

fi

echo "Finished"
